spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteDataFilesAction.java

```diff
@@ -1064,27 +1064,37 @@ protected void shouldHaveACleanCache(Table table) {
     Assert.assertNotEquals("Found no overlapping files", Collections.emptyList(), overlappingFiles);
   }
 
+  private <T> Pair<T, T> boundsOf(DataFile file, NestedField field, Class<T> javaClass) {
+    int columnId = field.fieldId();
+    return Pair.of(javaClass.cast(Conversions.fromByteBuffer(field.type(), file.lowerBounds().get(columnId))),
+        javaClass.cast(Conversions.fromByteBuffer(field.type(), file.upperBounds().get(columnId))));
+  }
+
+
   private <T> List<Pair<Pair<T, T>, Pair<T, T>>> getOverlappingFiles(Table table, String column) {
     table.refresh();
     NestedField field = table.schema().caseInsensitiveFindField(column);
-    int columnId = field.fieldId();
     Class<T> javaClass = (Class<T>) field.type().typeId().javaClass();
+
     Map<StructLike, List<DataFile>> filesByPartition = Streams.stream(table.currentSnapshot().addedFiles())
         .collect(Collectors.groupingBy(DataFile::partition));
 
     Stream<Pair<Pair<T, T>, Pair<T, T>>> overlaps =
         filesByPartition.entrySet().stream().flatMap(entry -> {
-          List<Pair<T, T>> columnBounds =
+          List<Pair<Pair<T, T>, Pair<T, T>>> boundsComparisons =
               entry.getValue().stream()
-                  .map(file -> Pair.of(
-                      javaClass.cast(Conversions.fromByteBuffer(field.type(), file.lowerBounds().get(columnId))),
-                      javaClass.cast(Conversions.fromByteBuffer(field.type(), file.upperBounds().get(columnId)))))
+                  .flatMap(left -> entry.getValue().stream()
```

Comment:
Maybe just my preference but it may be clearer with .zip?  , ie:

```
          Stream<DataFile> dataFiles = entry.getValue().stream();
          List<Pair<T, T>> columnBounds =
              Streams.zip(dataFiles, dataFiles, (left, right) -> Pair.of(left, right))
                  .filter(p -> p.first() != p.second())
                  .map(p -> Pair.of(boundsOf(p.first(), field, javaClass), boundsOf(p.second(), field, javaClass)))
                  .collect(Collectors.toList());
```

Reply:
Oh that's nice, I usually assume we don't have nice things in Java but of course this is in Guava :)

Reply:
One issue here is we need to do a foreach here, I'm not sure how that's working in your sample?

Reply:
Functional programming in java is just not that pretty. Let me make this iterative and it should be much clearer

Reply:
It's so sad because functional programming can be quite pretty. Every once in a while I try it in Java and it just makes me upset.